diff --git a/target/riscv/cpu_helper.c b/target/riscv/cpu_helper.c
index 9b821a971..2a4dd6188 100644
--- a/target/riscv/cpu_helper.c
+++ b/target/riscv/cpu_helper.c
@@ -978,8 +978,9 @@ restart:
             return TRANSLATE_FAIL;
         } else {
             /* if necessary, set accessed and dirty bits. */
-            target_ulong updated_pte = pte | PTE_A |
-                (access_type == MMU_DATA_STORE ? PTE_D : 0);
+            /* By MaPl: prevent accessed and dirty bits change in debug */
+            target_ulong updated_pte = pte | (!is_debug ? PTE_A : 0) |
+                (((access_type == MMU_DATA_STORE) && !is_debug) ? PTE_D : 0);
 
             /* Page table updates need to be atomic with MTTCG enabled */
             if (updated_pte != pte) {
