diff --git a/target/riscv/cpu_helper.c b/target/riscv/cpu_helper.c
index 1c60fb2e8..9b821a971 100644
--- a/target/riscv/cpu_helper.c
+++ b/target/riscv/cpu_helper.c
@@ -960,7 +960,7 @@ restart:
             /* User PTE flags when not U mode and mstatus.SUM is not set,
                or the access type is an instruction fetch */
             return TRANSLATE_FAIL;
-        } else if (!(pte & PTE_U) && (mode != PRV_S)) {
+        } else if (!(pte & PTE_U) && (mode != PRV_S) && !is_debug) {
             /* Supervisor PTE flags when not S mode */
             return TRANSLATE_FAIL;
         } else if (ppn & ((1ULL << ptshift) - 1)) {
